js-core-3/week-1/lesson.md

```diff
@@ -1,386 +1,267 @@
-> really zero? Run the tests using Jest.
+You should very clearly be able to say exact what you expected to happen when you run your code. If you cannot do this, then you do not understand the problem well enough.
 
-> _Exercise:_ Take the `findNeedle` function you have tested previously, copy it
-> into the `jest/` folder and call it `findNeedle.test.js`. Then write a test to
-> be used with Jest, similar to `sum.test.js`. Make sure you cover multiple
-> inputs and give all tests meaningful descriptions! Run the tests using Jest.
+- Is it actually broken?
 
-## Test Driven Development
+Sometime what you think is broken is actually working fine but not what you thought was going to happen. You should confirm with what the program is trying to do.
 
-Test-driven development (TDD) is a software development process that relies on the repetition of a very short development cycle: requirements are turned into very specific test cases, then the software is improved to pass the new tests, only. This is opposed to software development that allows software to be added that is not proven to meet requirements. [Wikipedia]
+- What happened instead?
 
-When developing following TDD, you normally follow this sequence:
+You should be able to quickly state what you were expecting to happen. Just saying "It Didn't Work" isn't enough! The more carefully you dissect the problem the more easily you'll be able to fix it.
 
-1. Add a test
-2. Run all tests and see if the new test fails (Red)
-3. Write the simplest code to make the test pass (Green)
-4. Refactor
-5. Repeat
+- What have I tried so far?
 
+Debugging is a process of loops and iteration. Think about what you've done so far and work out other ways that the code could have broken.
 
-Read more on the [Wikipedia article](https://en.wikipedia.org/wiki/Test-driven_development) and the resources at the end.
+#### Tools
 
-[Red Green Refactor](https://www.codecademy.com/articles/tdd-red-green-refactor)
+**Task (5 Minutes)**
 
-> Exercise: Two mentors pair on a problem doing "ping pong" TDD. One writing the test, the other writing the implementation.
+In teams, the students should have quick discussion about the ways that they have solved bugs or errors they've had in the past.
 
-## More on Testing
+When they finish the discussion we should regroup and gather together all the ways that they have solved bugs in the past.
 
-### Test coverage
+##### Thinking like a computer
 
-Test coverage describes the extent to which a code base is tested. When Jest
-runs your tests, it generates a so-called _coverage report_. This report tells
-you how many of your _lines of code_ are covered by tests, how many functions,
-statements, and branches.
+The most important skill you can develop is thinking analytically and logically - exactly the same as a computer. In the exercises soon we will be using our logical skills to slowly step through a program to see what is going wrong.
```

Comment:
```suggestion
The most important skill you can develop is thinking analytically and logically - exactly the same as a computer. In the exercises soon we will be using our logical skills to slowly step through a program to see what is going wrong.

Remember that the computer will run your program line-by-line, from top to bottom, and that loops and functions may jump to a new location. If in doubt, try going through the program one line at a time - don't jumping ahead, or guess what you _want_ your program to do, make sure you're reading what it actually _does_ do.

When debugging, it can often be useful to write down what values variables have, either on paper, in comments, or by using `console.log` - this is particularly true in loops and functions, where variable values may change from iteration to iteration, or between calls!
```